// node-instagram/lib/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var lodash_1 = require("lodash");
var node_fetch_1 = require("node-fetch");
var queryString = require("query-string");
var stream_1 = require("./stream");
var Instagram = (function () {
    function Instagram(config) {
        var apiVersion = config.apiVersion || 'v1';
        this.baseApiUrl = 'https://api.instagram.com';
        this.apiUrl = this.baseApiUrl + "/" + apiVersion + "/";
        this.config = {
            clientId: config.clientId,
            clientSecret: config.clientSecret,
            accessToken: config.accessToken,
        };
    }
    Instagram.prototype.get = function (endpoint, options, callback) {
        return this.request('GET', endpoint, options, callback);
    };
    Instagram.prototype.post = function (endpoint, options, callback) {
        return this.request('POST', endpoint, options, callback);
    };
    Instagram.prototype.delete = function (endpoint, options, callback) {
        return this.request('DELETE', endpoint, options, callback);
    };
    Instagram.prototype.stream = function (endpoint, options) {
        return new stream_1.default(this, endpoint, options);
    };
    Instagram.prototype.getAuthorizationUrl = function (redirectUri, options) {
        if (options === void 0) { options = {}; }
        var authorizationUrl = this.baseApiUrl + "/oauth/authorize/?client_id=" + this
            .config.clientId + "&redirect_uri=" + redirectUri + "&response_type=code";
        if (options.scope) {
            if (lodash_1.isArray(options.scope)) {
                options.scope = options.scope.join('+');
            }
            authorizationUrl += "&scope=" + options.scope;
        }
        if (options.state) {
            authorizationUrl += "&state=" + options.state;
        }
        return authorizationUrl;
    };
    Instagram.prototype.authorizeUser = function (code, redirectUri, callback) {
        return this.request('POST', 'oauth/access_token', {
            uriAbsolute: true,
            code: code,
            redirect_uri: redirectUri,
            client_id: this.config.clientId,
            client_secret: this.config.clientSecret,
            grant_type: 'authorization_code',
        }, callback);
    };
    Instagram.prototype.request = function (type, endpoint, options, callback) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var stringify, headers, uri, response, json, err_1, error;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        if (lodash_1.isFunction(options)) {
                            callback = options;
                            options = {};
                        }
                        stringify = true;
                        headers = {};
                        uri = "" + this.apiUrl + endpoint;
                        if (this.config.accessToken) {
                            options.access_token = this.config.accessToken;
                        }
                        if (options.accessToken) {
                            options.access_token = options.accessToken;
                            delete options.accessToken;
                        }
                        if (options.uriAbsolute) {
                            uri = this.baseApiUrl + "/" + endpoint;
                            delete options.uriAbsolute;
                        }
                        if (type === 'GET' || type === 'DELETE') {
                            uri += "?" + queryString.stringify(options);
                            options = null;
                        }
                        else if (type === 'POST') {
                            headers['Content-Type'] =
                                'application/x-www-form-urlencoded;charset=UTF-8';
                            options = queryString.stringify(options);
                            stringify = false;
                        }
                        return [4, node_fetch_1.default(uri, {
                                method: type,
                                body: options && stringify ? JSON.stringify(options) : options || null,
                                headers: headers,
                            })];
                    case 1:
                        response = _a.sent();
                        return [4, response.json()];
                    case 2:
                        json = _a.sent();
                        if (!response.ok) {
                            throw json;
                        }
                        if (lodash_1.isFunction(callback)) {
                            callback(null, json);
                        }
                        return [2, json];
                    case 3:
                        err_1 = _a.sent();
                        error = err_1.error || err_1;
                        if (lodash_1.isFunction(callback)) {
                            return [2, callback(error)];
                        }
                        throw error;
                    case 4: return [2];
                }
            });
        });
    };
    return Instagram;
}());
exports.default = Instagram;
